### DISCUSSION TITLE
Using zod to validate/parse an incoming property of an object but omit the property from output

### DISCUSSION THREAD

**Discussion by tanmayaBiswalOdiware** [colinhacks/zod#1717] [Q&A] [answered]
If this is my property:

```js
const questionObjectShape = z.object({
      title: z.string(),
      explanation: z.string(),
      answer: z.array(z.string).length(2)
})

const questionObj = {
      title: "A title",
      explanation: "An explanation"
      answer: ["huhu", "haha"]
}

const result = questionObjectShape.parse(questionObj)
```

What I want here is to validate the answer property but when I get a success, it returns an object without the answer property. But when the answer.length is less than 2, it throws an error.

From what I had tried earlier, using omit actually ignores the omitted property altogether. So what can I do here?

**Comment by JacobWeisenburger** [comment]
Is this what you are looking for?

https://github.com/colinhacks/zod#transform

**Reply by tanmayaBiswalOdiware** [reply]
So you are saying that I transform the input to delete one of the property and return it?

**Reply by JacobWeisenburger** [reply]
Yes. That's one way to do it

**Comment by JacobWeisenburger** [comment, marked as answer]
Here's a full solution if you need more details:

```ts
const questionObjectShape = z.object( {
    title: z.string(),
    explanation: z.string(),
    answer: z.string().array().length( 2 )
} )
type QuestionObjectShape = z.infer<typeof questionObjectShape>
// type QuestionObjectShape = {
//     title: string
//     explanation: string
//     answer: string[]
// }

const transformedQuestionObjectShape = questionObjectShape.transform( value =>
    Object.fromEntries(
        Object.entries( value ).filter( ( [ key ] ) => key !== 'answer' )
    ) as Omit<QuestionObjectShape, 'answer'>
)
type TransformedQuestionObjectShape = z.infer<typeof transformedQuestionObjectShape>
// type TransformedQuestionObjectShape = {
//     title: string
//     explanation: string
// }

console.log(
    questionObjectShape.parse( {
        title: 'A title',
        explanation: 'An explanation',
        answer: [ 'foo', 'bar' ],
    } )
) // { title: 'A title', explanation: 'An explanation', answer: [ 'foo', 'bar' ] }

console.log(
    transformedQuestionObjectShape.parse( {
        title: 'A title',
        explanation: 'An explanation',
        answer: [ 'foo', 'bar' ],
    } )
) // { title: 'A title', explanation: 'An explanation' }

const failedParse = transformedQuestionObjectShape.safeParse( {
    title: 'A title',
    explanation: 'An explanation',
    answer: [ 'foo' ],
} )
if ( !failedParse.success ) console.log(
    failedParse.error.issues[ 0 ]
)
// {
//     code: "too_small",
//     minimum: 2,
//     maximum: undefined,
//     type: "array",
//     inclusive: true,
//     exact: true,
//     message: "Array must contain exactly 2 element(s)",
//     path: [ "answer" ]
// }
```

**Comment by tanmayaBiswalOdiware** [comment]
Thanks on elaborating with typescript annotations!

I figured it out when you told me about transform which also made me realize that I can also change a property name from inside transform as well. Which is cool. I had that doubt too.

But I have another small doubt too, if you don't mind me asking it here - 

If I am validating an array of object, and I am returned a ZodError with `path`, say: 

```ts
"path": [
      3,
      "type"
],
```

where `3` is the index value of the array where error is found. `type` being the property which was invalidated.

Can I write custom error for this? Or do I just catch the error and  process it myself?

**Reply by JacobWeisenburger** [reply]
This seems like an unrelated question. Would you be able to put it into a new discussion? I'm more than happy to answer it there. Sorry for the hassle, I'm just wanting other people to be able to find that answer easily in the future. Thanks

**Reply by tanmayaBiswalOdiware** [reply]
Sure sure, I just thought you might prefer to end the discussion in this thread without spamming the discussion with my questions lol. But it makes sense you want everyone to get as much answer as they can. I will post the question after framing it properly. No worries. Have a good day.

**Reply by JacobWeisenburger** [reply]
Perhaps this would help: https://github.com/colinhacks/zod/blob/master/ERROR_HANDLING.md
